cmd/zdb/zdb.c

```diff
@@ -3631,6 +3631,9 @@ parse_object_range(char *range, zopt_object_range_t *zor, char **msg)
 			*msg = "Invalid characters in object ID";
 			rc = 1;
 		}
+#ifdef __APPLE__
+		zor->zor_obj_start = INO_XNUTOZFS(zor->zor_obj_start, 2);
+#endif
```

Comment:
Out of curiosity where exactly do 0-15 get mapped?  What are they used for?  You've probably noticed we do something similar for the `.zfs/` directories on Linux with using high numbered object ids (although those never get written to disk).

I'd suggest we add a wrapper macro for this to get rid of the `#ifdef __APPLE` blocks.  Maybe something private to `zdb` since it's only needed here like:

```c
#ifdef __APPLE__
#define ZDB_MAP_OBJECT_ID(obj, num) INO_XNUTOZFS(obj, num)
#else
#define ZDB_MAP_OBJECT_ID(obj, num) (obj)
#endif
```

Reply:
Ah my apologies - I do try to show the macOS side of the code in the PRs, but forgot this time.
```
#define INO_ROOT                2ULL
#define INO_RESERVED            16ULL   /* [0-15] reserved. */
#define INO_ISRESERVED(ID)      ((ID) < (INO_RESERVED))
/*                              0xFFFFFFFFFFFFFFF0 */
#define INO_MAP                 ((uint64_t)-INO_RESERVED) /* -16, -15, .., -1 */

#define INO_ZFSTOXNU(ID, ROOT)  \
        ((ID) == (ROOT)?INO_ROOT:(INO_ISRESERVED(ID)?INO_MAP+(ID):(ID)))
```
So they go to 0xFFFFFFFFFFFFFFF0 - 0xFFFFFFFFFFFFFFFF
(and the ID of the mountroot is always `2`, also part of the map)

Which you can clearly see with `ls -li` on a newly created dir:

```
                   2 drwxr-xr-x  18 lundman  wheel         21 Aug 31 15:38 .
         12884925458 drwxr-xr-x   5 root     wheel        160 Sep  1 11:07 ..
18446744073709551602 drwx------   3 lundman  wheel          4 Aug 16 09:42 .Spotlight-V100
18446744073709551607 d-wx--x--t   3 root     wheel          3 Aug 16 09:42 .Trashes
18446744073709551606 drwx------   2 lundman  wheel        268 Sep  1 09:39 .fseventsd
                   2 drwxr-xr-x   9 lundman  wheel         10 Aug 16 10:08 4tf
               52924 drwxr-xr-x  10 502      wheel         20 Aug 16 12:25 go
```
Since `.fseventsd`, `.Spotlight-V100`, and `.Trashes` are always created immediately on a new filesystem they always end up in the 0-15 range, and then end up breaking spotlight (and/or Trashcan).

If you are curious why HFS takes the range:
```
        kHFSRootParentID                = 1,    /* Parent ID of the root folder 
*/
        kHFSRootFolderID                = 2,    /* Folder ID of the root folder */
        kHFSExtentsFileID               = 3,    /* File ID of the extents file */
        kHFSCatalogFileID               = 4,    /* File ID of the catalog file */
        kHFSBadBlockFileID              = 5,    /* File ID of the bad allocation block file */
        kHFSAllocationFileID            = 6,    /* File ID of the allocation file (HFS Plus only) */
        kHFSStartupFileID               = 7,    /* File ID of the startup file (HFS Plus only) */
        kHFSAttributesFileID            = 8,    /* File ID of the attribute file (HFS Plus only) */
        kHFSAttributeDataFileID         = 13,   /* Used in Mac OS X runtime for extent based attributes */
                                                /* kHFSAttributeDataFileID is never stored on disk. */
        kHFSRepairCatalogFileID         = 14,   /* Used when rebuilding Catalog B-tree */
        kHFSBogusExtentFileID           = 15,   /* Used for exchanging extents in extents file */
        kHFSFirstUserCatalogNodeID      = 16
```

I still am not sure if this is a limit with `apfs`. 


Adding the empty MAP macro next push.

Reply:
Oh and worth pointing out that these mapping also do not get written to disk. If a Linux pool uses ID 15, it is presented to XNU as `0xffffffffffffffff`, and reversed in `zfs_vget()`.  
(We do the same ID mapping in .zfs for the ".zfs/snapshot/*" entries like Linux, but for us, everything is a znode_t, even those fakes directories - don't know if Linux does that too, but old gfs.c used to create own "not-znode_t" structs)

`zdb` doesn't "need" the mapping, it is simply for the User, if they `ls -li` and paste the "large ID" into zdb, it won't find it (as they'd need to map it). Not sure why we were being nice to the users that day.

Reply:
>  Not sure why we were being nice to the users that day

To keep them guessing of course! Or perhaps for your own sanity when debugging.